server/matcher/match.go

```diff
@@ -0,0 +1,419 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package matcher
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"math/rand"
+	"sort"
+
+	"github.com/decred/dcrd/crypto/blake256"
+	"github.com/decred/dcrdex/server/market/order"
+)
+
+// HashFunc is the hash function used to generate the shuffling seed.
+var HashFunc = blake256.Sum256
+
+const (
+	HashSize = blake256.Size
+)
+
+type Matcher struct {
+	marketBuyBuffer uint64 // base asset amount required above lot size for market buys
```

Comment:
Actually, I just realized the market buy buffer check should to be done when validating incoming orders, rather than starting the matching for an order.  So I think it's a mistake to even do the check when matching starts because even if the rate only matches exactly it's still fine since the rate of the matched order is known at that point.  The buffer only has a purpose before the exact rate is known.

But regarding the units of the buffer, I'm not sure I follow "the buffer becomes reactive to market shifts".  Presently I have:

```go
func (m *Matcher) CheckMarketBuyBuffer(book Booker, ord *order.MarketOrder) bool {
	rate := book.BestSell().Rate
	minBase := m.marketBuyBuffer + book.LotSize()
	return ord.Remaining() >= uint64(rate*float64(minBase))
}
```

vs.

```go
	minBase := m.marketBuyBuffer * book.LotSize()
	return ord.Remaining() >= uint64(rate*float64(minBase))
```

They are both fine since the lot size is constant, and rate is still multiplied.  Sorry if I've misunderstood your comment.

Reply:
Yeah. That wasn't very clear on my part. I guess I'm just saying that by setting it as a ratio of lot size, the operator doesn't need to consider it if the market shifts and the lot size needs to be adjusted. 

Example: for some asset XYZ, lot size is set to 10 XYZ, buffer is set to 5 XYZ. In the `l + b` sheme b=5, but in the `l * b` scheme b=0.5. Then XYZ gains value and the operator chooses to change the lot size to 5 XYZ. In the `l + b` scheme, they would likely adjust the buffer so that b=2.5. That conversion is automatic in the `l * b` scheme. 

--

Agree that it should be handled during pre-processing.

Reply:
Oh, right makes sense it's better as a ratio for when lot size needs to change.
As for the client needing to know the best standing sell order rate, that would be required in either the multiplicative or additive buffers, at least in how I've defined it.
Switching to float...

Reply:
Resolved in https://github.com/decred/dcrdex/pull/15/commits/9e62c61694f285db2c39489118e2a5b466a9c1e5 by removing the market buy buffer from the matcher, although when we make order validation we'll use float as defined in the spec.